db/migrate/20230803225825_container_indexes.rb

```diff
@@ -0,0 +1,25 @@
+class ContainerIndexes < ActiveRecord::Migration[6.1]
+  def change
+    # container projects page
+    add_index :container_routes, :container_project_id
+    add_index :container_services, :container_project_id
+    # didn't show much, since only had 1 record
+    add_index :container_replicators, :container_project_id
+
+    add_index :container_groups, [:container_project_id, :id], :name => "index_container_groups_on_cpid_id_not_del", :where => "deleted_on IS NULL"
+    # container providers screen
+
+    # NOTE: this gets index only scans, but has a larger index size
+    add_index :taggings, [:taggable_type, :taggable_id, :tag_id], :name => "index_taggings_on_type_id_id"
+    remove_index :taggings, :column => [:taggable_id, :taggable_type], :name => "index_taggings_on_taggable_id_and_taggable_type"
+
+    # container Services
+    add_index :container_groups_container_services, [:container_service_id, :container_group_id], :name => "index_container_groups_on_csi_cgi"
+
+    # container groups
+    add_index :containers, [:container_group_id, :state]
+
+    # container nodes
+    add_index :container_conditions, [:container_entity_type, :container_entity_id, :name], :name => "index_container_conditions_on_cet_ceid_name"
```

Comment:
I would expect id first.

Reply:
The virtual attribute that this is targeting is:

```
  (SELECT container_conditions.status
  FROM container_conditions
  WHERE container_conditions.name = 'Ready'
  AND container_conditions.container_entity_id = container_nodes.id
  AND container_conditions.container_entity_type = 'ContainerNode')
  AS ready_condition_status,
```

When we have 2 indexes, one with `id` first and the other with `type` first, the analyzer prefers the one with `id` first. When I delete the `id` first one, it is still able to use the `type` first one, and the query stats are similar.

Having `name, type, id` or `WHERE name = 'READY'` would be faster, but this way, the index can also be used for all `container_conditions`, not just the `ready` ones.

Reply:
> When we have 2 indexes, one with id first and the other with type first, the analyzer prefers the one with type first. When I delete the type first index, it is still able to use the id first index with similar query statistics.

I'm not sure that's an entirely accurate assessment.  Or rather, statistically, if they are the same, it will just pick one, but that doesn't then imply that type,id is better than id,type.

Based on all my readings, higher cardinality should always be first because it eliminates more rows and allows the index to execute faster, because the row elimination happens on the first pass of the index.  So id,type is nearly always preferable over type,id. The only time this shouldn't be true is if we have queries that don't provider the _id in the query at all.

> Having `name, type, id` ...

Not my suggestion - I think it should be `id, type, name`